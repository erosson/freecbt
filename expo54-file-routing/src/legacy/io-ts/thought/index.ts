export { Thought, ID, create, Group, groupByDay } from "./thought"
export { Persist, Legacy, VERSION } from "./persist"
export { Codec, FromLegacy } from "./codec"
