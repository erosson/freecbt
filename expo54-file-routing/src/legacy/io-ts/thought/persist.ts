import * as T from "io-ts"
import * as Distortion from "../distortion"

export const VERSION = "Thought-v1"

/**
 * our json-formatted thought data, as persisted to disk
 */
export const Persist = T.type(
  {
    v: T.literal(VERSION),
    automaticThought: T.string,
    alternativeThought: T.string,
    cognitiveDistortions: T.array(T.string),
    challenge: T.string,
    createdAt: T.string,
    updatedAt: T.string,
    uuid: T.string,
  },
  "Thought.Persist"
)
export type Persist = T.TypeOf<typeof Persist>

/**
 * old-style thoughts, as json
 * These were persisted in user data, so we must maintain decode support forever
 */
export const Legacy = T.intersection(
  [
    T.type({
      automaticThought: T.string,
      alternativeThought: T.string,
      cognitiveDistortions: T.array(Distortion.LegacyID),
      challenge: T.string,
      createdAt: T.string,
      updatedAt: T.string,
      uuid: T.string,
    }),
    T.partial({ v: T.undefined }),
  ],
  "Thought.Legacy"
)
export type Legacy = T.TypeOf<typeof Legacy>
