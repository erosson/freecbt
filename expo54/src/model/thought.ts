import { v4 as uuidv4 } from "uuid";
import { z } from "zod";
import * as Distortion from "./distortion";

export const Json = z.object({
  automaticThought: z.string(),
  cognitiveDistortions: z.string().array(),
  challenge: z.string(),
  alternativeThought: z.string(),
  createdAt: z.iso.datetime(),
  updatedAt: z.iso.datetime(),
  uuid: z.string(),
});
export type Json = z.infer<typeof Json>;

export const ID_PREFIX = `@Quirk:thoughts:`;
export const Thought = z.object({
  automaticThought: z.string(),
  cognitiveDistortions: z.set(Distortion.Distortion),
  challenge: z.string(),
  alternativeThought: z.string(),
  createdAt: z.date(),
  updatedAt: z.date(),
  uuid: z.string(),
});
export type Thought = z.infer<typeof Thought>;

export function isKey(key: string): boolean {
  return key.startsWith(ID_PREFIX);
}

export type Spec = Pick<
  Thought,
  | "automaticThought"
  | "cognitiveDistortions"
  | "challenge"
  | "alternativeThought"
>;
export function create(spec: Spec, now: Date): Thought {
  const createdAt = now;
  const updatedAt = now;
  const uuid = `${ID_PREFIX}${uuidv4()}`;
  return { ...spec, createdAt, updatedAt, uuid };
}

/**
 * Thought-parsing enforces valid distortions, but parameterize the distortion-data until later.
 */
export function createParsers(data: Distortion.Data) {
  const { fromSlugSet: distortionsFromSlugSet } =
    Distortion.createParsers(data);

  const fromJson = z.codec(Json, Thought, {
    decode: (json: Json) => {
      const cognitiveDistortions = distortionsFromSlugSet.decode(
        new Set(json.cognitiveDistortions)
      );
      const createdAt = new Date(json.createdAt);
      const updatedAt = new Date(json.updatedAt);
      return Thought.decode({
        ...json,
        cognitiveDistortions,
        createdAt,
        updatedAt,
      });
    },
    encode: (t: z.input<typeof Thought>) => {
      const cognitiveDistortions = Array.from(t.cognitiveDistortions).map(
        (d) => d.slug
      );
      const createdAt = t.createdAt.toISOString();
      const updatedAt = t.updatedAt.toISOString();
      return Json.decode({ ...t, cognitiveDistortions, createdAt, updatedAt });
    },
  });

  const fromString = z.codec(z.string(), fromJson, {
    decode: (enc: string) => JSON.parse(enc),
    encode: (dec: Json) => JSON.stringify(dec),
  });
  return { fromJson, fromString };
}
