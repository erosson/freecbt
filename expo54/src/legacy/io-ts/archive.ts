import * as T from "io-ts"
import * as Thought from "./thought"
import * as LZ from "lz-string"
import { JsonFromString } from "io-ts-types"
import { pipe } from "fp-ts/lib/function"

const VERSION = "Archive-v1"

export const Archive = T.type(
  {
    v: T.literal(VERSION),
    thoughts: T.array(Thought.Persist),
  },
  "Archive.Raw"
)
export type Archive = T.TypeOf<typeof Archive>

export function create(thoughts: Thought.Persist[]): Archive {
  return { v: VERSION, thoughts }
}

const LZStringBase64 = T.string.pipe(
  new T.Type(
    "LZString",
    T.string.is,
    (input: string, c) => {
      const ret = LZ.decompressFromBase64(input)
      // decompress returns "" on error. But "" is valid output too, since we
      // can compress ""! Special-case that one.
      return ret === null || (ret === "" && input !== "Q===")
        ? T.failure("Invalid lz-string encoding", c)
        : T.success(ret)
    },
    LZ.compressToBase64
  )
)

function FromPrefix(
  pre: string,
  message: string
): T.Type<string, string, string> {
  return new T.Type(
    `FromPrefix(${pre})`,
    T.string.is,
    (enc: string, c) =>
      enc.startsWith(pre)
        ? T.success(enc.slice(pre.length))
        : T.failure(message, c),
    (dec: string) => `${pre}${dec}`
  )
}

function FromSuffix(
  suf: string,
  message: string
): T.Type<string, string, string> {
  return new T.Type(
    `FromSuffix(${suf})`,
    T.string.is,
    (enc: string, c) =>
      enc.endsWith(suf)
        ? T.success(enc.slice(0, -suf.length))
        : T.failure(message, c),
    (dec: string) => `${dec}${suf}`
  )
}

// A FreeCBT archive string is a prefix, followed by an LZString-base64-compressed json body
export const Codec: T.Type<Archive, string> = T.string
  .pipe(FromPrefix(":FreeCBT:", "Not a FreeCBT archive"))
  .pipe(FromSuffix(":FreeCBT:", "Not a FreeCBT archive"))
  .pipe(LZStringBase64)
  .pipe(JsonFromString)
  .pipe(Archive, "Archive.Codec")
