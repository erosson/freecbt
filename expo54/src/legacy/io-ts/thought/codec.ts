import { Persist, Legacy, VERSION } from "./persist"
import { Thought } from "./thought"
import * as T from "io-ts"
import { pipe } from "fp-ts/lib/function"
import * as Distortion from "../distortion"

const FromPersist = Persist.pipe(Thought)

// encoding legacy-thoughts is still useful for testing
export const FromLegacy = Legacy.pipe(
  new T.Type(
    "ThoughtFromLegacy",
    Thought.is,
    // the distortion codec understands the legacy format, no need to transform them here
    (leg: Legacy, c) => Thought.validate({ ...leg, v: VERSION }, c),
    (dis: Thought): Legacy => ({
      ...FromPersist.encode(dis),
      v: undefined,
      cognitiveDistortions: pipe(
        dis.cognitiveDistortions,
        Array.from,
        T.array(Distortion.FromLegacy).encode
      ),
    })
  )
)

export const Codec: typeof FromPersist = new T.Type(
  "Thought.Codec",
  FromPersist.is,
  T.union([FromPersist, FromLegacy]).validate,
  FromPersist.encode
)
