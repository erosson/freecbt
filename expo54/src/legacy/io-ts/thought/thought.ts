import * as T from "io-ts"
import { setFromArray, DateFromISOString } from "io-ts-types"
import * as Distortion from "../distortion"
// import this polyfill before `uuid`: https://www.npmjs.com/package/uuid#user-content-getrandomvalues-not-supported
import "react-native-get-random-values"
import { v4 as uuidv4 } from "uuid"
import { VERSION } from "./persist"

export const ID = T.string
export type ID = T.TypeOf<typeof ID>

export const Thought = T.type(
  {
    v: T.literal(VERSION),
    automaticThought: T.string,
    alternativeThought: T.string,
    cognitiveDistortions: Distortion.SetCodec,
    challenge: T.string,
    createdAt: DateFromISOString,
    updatedAt: DateFromISOString,
    uuid: ID,
  },
  "Thought"
)
export type Thought = T.TypeOf<typeof Thought>

export interface CreateArgs {
  automaticThought: string
  alternativeThought: string
  cognitiveDistortions: Iterable<Distortion.Distortion | string>
  challenge: string
  createdAt?: Date
  updatedAt?: Date
  uuid?: string
}
export const THOUGHTS_KEY_PREFIX = `@Quirk:thoughts:`
export function getThoughtKey(info: string): string {
  return `${THOUGHTS_KEY_PREFIX}${info}`
}

export function create(args: CreateArgs): Thought {
  const cognitiveDistortions = new Set(
    Array.from(args.cognitiveDistortions).map((name) => {
      const d = typeof name === "string" ? Distortion.bySlug[name] : name
      if (!d) {
        throw new Error(`no such distortion: ${name}`)
      }
      return d
    })
  )
  const uuid = args.uuid ?? getThoughtKey(uuidv4())
  return {
    ...args,
    cognitiveDistortions,
    uuid,
    createdAt: args.createdAt ?? new Date(),
    updatedAt: args.updatedAt ?? new Date(),
    v: VERSION,
  }
}

export interface Group {
  date: string
  thoughts: Thought[]
}

export function groupByDay(thoughts: Thought[]): Group[] {
  const dates: string[] = []
  const groups: Group[] = []

  const sortedThoughts = thoughts.sort(
    (first, second) => second.createdAt.getTime() - first.createdAt.getTime()
  )

  for (const thought of sortedThoughts) {
    const date = thought.createdAt.toDateString()
    if (!dates.includes(date)) {
      dates.push(date)
      groups.push({
        date,
        thoughts: [thought],
      })
      continue
    }

    groups[dates.length - 1].thoughts.push(thought)
  }

  return groups
}
