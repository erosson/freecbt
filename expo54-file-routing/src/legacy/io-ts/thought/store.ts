import AsyncStorage from "@react-native-async-storage/async-storage"
import * as AsyncState from "../../async-state"
import * as T from "io-ts"
import { JsonFromString } from "io-ts-types"
import * as E from "fp-ts/lib/Either"
import { Thought, ID, THOUGHTS_KEY_PREFIX } from "./thought"
import { Persist } from "./persist"
import { Codec } from "./codec"
import { decodeOrThrow } from "../io-utils"
import * as Archive from "../archive"
import { KeyValuePair } from "@react-native-async-storage/async-storage/lib/typescript/types"
import { pipe } from "fp-ts/lib/function"
import { scryRenderedComponentsWithType } from "react-dom/test-utils"

const EXISTING_USER_KEY = "@Quirk:existing-user"

export async function exists(key: ID): Promise<boolean> {
  try {
    const value = await AsyncStorage.getItem(key)
    return !!value
  } catch (err) {
    console.error(err)
    return false
  }
}

export async function getIsExistingUser(): Promise<boolean> {
  try {
    const value = await AsyncStorage.getItem(EXISTING_USER_KEY)
    return !!value
  } catch (err) {
    console.error(err)
    return false
  }
}

export async function setIsExistingUser() {
  try {
    await AsyncStorage.setItem(EXISTING_USER_KEY, "true")
  } catch (err) {
    console.error(err)
  }
}

export async function write(t: Thought): Promise<void> {
  const enc = Codec.encode(t)
  const raw = JSON.stringify(enc)
  await AsyncStorage.setItem(t.uuid, raw)
}

export async function read(id: ID): Promise<Thought> {
  return parseOrThrow((await AsyncStorage.getItem(id)) ?? "")
}
export async function readResult(
  id: ID
): Promise<AsyncState.Result<Thought, ParseError>> {
  return parseResult((await AsyncStorage.getItem(id)) ?? "", id)
}

export interface ParseError {
  error: any
  raw: string
  id: ID
}
function parseOrThrow(raw: string): Thought {
  const enc = JSON.parse(raw)
  return decodeOrThrow(Codec, enc)
}
function parseResult(
  raw: string,
  id: ID
): AsyncState.Result<Thought, ParseError> {
  return AsyncState.mapError(
    AsyncState.tryResult(() => parseOrThrow(raw)),
    (error) => ({ error, raw, id })
  )
}

export async function remove(uuid: ID) {
  try {
    await AsyncStorage.removeItem(uuid)
  } catch (error) {
    console.error(error)
  }
}

export async function getExercisesKeys(): Promise<string[]> {
  const allKeys = await AsyncStorage.getAllKeys()
  return allKeys.filter((key) => key.startsWith(THOUGHTS_KEY_PREFIX))
}
export async function getRawExercises(): Promise<readonly KeyValuePair[]> {
  const keys = await getExercisesKeys()
  return await AsyncStorage.multiGet(keys)
}
export async function getExercises(): Promise<
  AsyncState.Result<Thought, ParseError>[]
> {
  const rows = await getRawExercises()
  return rows.map(([key, raw]) => parseResult(raw ?? "", key))
}

export async function getValidExercises(): Promise<Thought[]> {
  const ts = await getExercises()
  return ts.filter(AsyncState.isSuccess).map((t) => t.value)
}

export const countThoughts = async (): Promise<number> => {
  const exercises = await getExercises()
  return exercises.length
}

export async function readArchive(): Promise<Archive.Archive> {
  const rows = await getRawExercises()
  const thoughts: Persist[] = rows
    .map(([key, raw]) => {
      const result = JsonFromString.pipe(Persist).decode(raw ?? "")
      return E.fold(
        (err) => [],
        (persist: Persist) => [persist]
      )(result)
    })
    .flat()
  return Archive.create(thoughts)
}

export async function writeArchive(archive: Archive.Archive): Promise<void> {
  const rows: [string, string][] = archive.thoughts.map((entry) => [
    entry.uuid,
    JsonFromString.pipe(Persist).encode(entry),
  ])
  const oldKeys = await getExercisesKeys()
  // asyncstorage gets upset about empty lists, even though there's nothing wrong with them in theory
  if (oldKeys.length) {
    await AsyncStorage.multiRemove(oldKeys)
  }
  if (rows.length) {
    await AsyncStorage.multiSet(rows)
  }
}

export async function readArchiveString(): Promise<string> {
  return pipe(await readArchive(), Archive.Codec.encode)
}

export async function writeArchiveString(
  enc: string
): Promise<T.Errors | null> {
  return pipe(
    enc.trim(),
    Archive.Codec.decode,
    E.fold(
      async (err: T.Errors): Promise<T.Errors | null> => err,
      async (dec: Archive.Archive) => {
        await writeArchive(dec)
        return null
      }
    )
  )
}
