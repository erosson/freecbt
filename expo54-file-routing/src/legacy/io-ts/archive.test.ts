import * as A from "./archive"
import * as E from "fp-ts/Either"
import * as T from "./thought"

const testdata: { name: string; dec: A.Archive }[] = [
  {
    name: "empty",
    dec: A.create([]),
  },
  {
    name: "nonempty",
    dec: A.create(
      [
        {
          automaticThought: "auto",
          challenge: "chal",
          alternativeThought: "alt",
          cognitiveDistortions: ["all-or-nothing"],
          createdAt: new Date(1234),
          updatedAt: new Date(1234),
          uuid: "someuuid",
        },
      ].map((args) => T.Codec.encode(T.create(args)))
    ),
  },
  {
    name: "multiple",
    dec: A.create(
      [
        {
          automaticThought: "auto",
          challenge: "chal",
          alternativeThought: "alt",
          cognitiveDistortions: ["all-or-nothing"],
          createdAt: new Date(1234),
          updatedAt: new Date(1234),
          uuid: "someuuid",
        },
        {
          automaticThought: "auto2",
          challenge: "chal2",
          alternativeThought: "alt2",
          cognitiveDistortions: ["all-or-nothing", "should-statements"],
          createdAt: new Date(4321),
          updatedAt: new Date(4321),
          uuid: "someuuid2",
        },
      ].map((args) => T.Codec.encode(T.create(args)))
    ),
  },
]

test.each(testdata)("encode/decode archives: $name", ({ dec }) => {
  const enc: string = A.Codec.encode(dec)
  expect(enc).toMatch(/^:FreeCBT:/)
  expect(enc).toMatchSnapshot()
  expect(A.Codec.encode(dec)).toEqual(enc)
  expect(A.Codec.decode(enc)).toEqual(E.right(dec))
  expect(A.Codec.decode("!" + enc)).toEqual(E.left(expect.anything()))
})
