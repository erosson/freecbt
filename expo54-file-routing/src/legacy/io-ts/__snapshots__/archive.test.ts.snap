// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`encode/decode archives: empty 1`] = `":FreeCBT:N4IgbiBcIIIE4GMAWBLMBTAtGAjCANCAC5ID2ArgOZJEDOUA2gLoC+QA:FreeCBT:"`;

exports[`encode/decode archives: multiple 1`] = `":FreeCBT:N4IgbiBcIIIE4GMAWBLMBTAtGAjCANCAC5ID2ArgOZJEDOUA2qAIblGkC2zRKCAKmSo0oIVuwIhkzADbT0AO0roRU6RJlF0cedzToBFakREaJCUpXkoeGACIpa7OD1Lz6kBqNmZScTPNISFEUQAF1CBDh0bnQAExhjaBwATgB2AAZMdJwsnD509MgCopwAOgAmAGYAFgAtCXIAB1iY+MSQFIzc3PzC4uyKmvrCcnIUWJFaTnRR8YkIaAMhImw8AF98FjZOXX5BIxNt8rMkGTlFZWhVY8INLR0bfX3haA0byQsrR/tHXxc3RheaQ+PwBIIhQi0QTSWKYRwxDgKOhhCJRVoJESdTLZHr9PrVUqVco4YYgJotTRtTFpbE5bK9Ir4wnE0mzCbQKaItnvBYgJZGVYgNahNZAA:FreeCBT:"`;

exports[`encode/decode archives: nonempty 1`] = `":FreeCBT:N4IgbiBcIIIE4GMAWBLMBTAtGAjCANCAC5ID2ArgOZJEDOUA2qAIblGkC2zRKCAKmSo0oIVuwIhkzADbT0AO0roRU6RJlF0cedzToBFakREaJCUpXkoeGACIpa7OD1Lz6kBqNmZScTPNISFEUQAF1CBDh0bnQAExhjaBwATgB2AAZMdJwsnD509MgCopwAOgAmAGYAFgAtCXIAB1iY+MSQFIzc3PzC4uyKmvrCcnIUWJFaTnRR8YkIaAMhImw8AF9QtaA===:FreeCBT:"`;
